### DISCUSSION TITLE
Filtered stream encountering HTTP error: 409

### DISCUSSION THREAD

**Discussion by 4RCAN3** [tweepy/tweepy#1945] [Questions] [answered]
I seem to be getting 409 error on my stream. From what I've researched, it seems like the error could be due to multiple instances of stream running. However, I don't have any other instances of stream running. I've tested this on multiple accounts and re-created multiple apps but I still seem to have this issue. Stream.sample() also seems to be working, however when I filter, I get the error. Can someone help out? 

```
class Stream(tweepy.StreamingClient):

    def on_connect(self):
        print('Stream started')

    def on_tweet(self, tweet):
        time.sleep(0.2)
        pass


stream = Stream(bearer_token=BEARER_TOKEN, wait_on_rate_limit=True)
stream.add_rules(tweepy.StreamRule(['@MyAccount is:reply']))
stream.filter(tweet_fields = ['conversation_id', 'referenced_tweets'])```

**Comment by dvdblk** [comment]
Literally the same issue here. I've used a bearer token and getting a `409`.

Best part about this is that Twitter doesn't even mention `409` on their [developer documentation site](https://developer.twitter.com/en/support/twitter-api/error-troubleshooting). ♿ 

EDIT: Even though it's not a fix it looks like the issue is in the `StreamingClient.filter()` method. `StreamingClient.sample()` works as expected.

**Comment by dvdblk** [comment]
While debugging I have found that the issue was caused by not adding `StreamRule`s properly.

```
{"title":"ConnectionException","detail":"You must define rules using the POST /2/tweets/search/stream/rules endpoint before connecting to filtered stream.","connection_issue":"RuleConfigurationIssue","type":"https://api.twitter.com/2/problems/streaming-connection"}
Stream encountered HTTP Error: 409
```

After adding an example rule from Twitter Docs the filter method doesn't result in a `409`.

**Working example**:
```
stream.add_rules(StreamRule("cat has:media"))
stream.filter()
```

**Comment by 4RCAN3** [comment]
That is correct. I figured a workaround by using twarc to add rules instead. I'll check out your method as well. It's very inconvenient that the error wasn't mentioned properly in the docs. But anyhow, I suppose this works.

**Comment by Harmon758** [comment, marked as answer]
[`StreamRule`](https://tweepy.readthedocs.io/en/v4.10.0/streamrule.html) expects a string.

`stream.add_rules(tweepy.StreamRule(['@MyAccount is:reply']))` will return a Response with an error:

> `Response(data=None, includes={}, errors=[{'value': '["@MyAccount is:reply"]', 'details': ["Cannot parse rule at '[' (position 1)"], 'title': 'UnprocessableEntity', 'type': 'https://api.twitter.com/2/problems/invalid-rules'}], meta={'sent': '[ . . .]', 'summary': {'created': 0, 'not_created': 1, 'valid': 0, 'invalid': 1}})`

I believe the 409 Conflict error occurs when you subsequently attempt to start the stream with no existing rules, since that rule was invalid and not created.

You can simply add the rule directly as a string instead, e.g. `stream.add_rules(tweepy.StreamRule('@MyAccount is:reply'))`.

As for the lack of documentation for 409 errors on Twitter's end, I've passed that feedback along.
